### ISSUE TITLE
[HOLD for payment 2023-07-21] [$1000] Message “Complete task…” should not be displayed when the user is marked as done task not assigned

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#20542]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Login 2 accounts on 2 devices A and B
2. A Create chat with B
3. A Create a task 1 but not assign anyone
4. B click into task 1
5. B click Mark as Done
## Expected Result: 
The message complete task shouldn’t show
## Actual Result:: 
The message complete task show

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.26-1
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/dfafad65-9b97-413a-a53f-5b6b3d89fc76

https://github.com/Expensify/App/assets/43996225/c2f2afc5-133d-4b98-8b46-99b72af47b7c

**Expensify/Expensify Issue URL:**
**Issue reported by:** @namhihi237
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1685993057669719

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~019928155ae5bbd4b8</li>
        <li>Upwork Job ID: 1668264032406409216</li>
        <li>Last Price Increase: 2023-06-27</li>
    </ul>
</details>

**Comment by BhuvaneshPatil** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The task that has no assignee can be marked as done by other user

### What is the root cause of that problem?
There are three places from where a task can be marked as completed
- A check box in TaskPreview
https://github.com/Expensify/App/blob/4cad2489f6feb5efb955924764711a0c36439f58/src/components/ReportActionItem/TaskPreview.js#L69-L81

- Mark As Done button in TaskHeader
https://github.com/Expensify/App/blob/4cad2489f6feb5efb955924764711a0c36439f58/src/components/TaskHeader.js#L94-L100

- Mark as Done option in HeaderView
https://github.com/Expensify/App/blob/4cad2489f6feb5efb955924764711a0c36439f58/src/pages/home/HeaderView.js#L92-L96

At all these places we are not adding any check if the button can be clicked by particular user or not.
These buttons are visible to everyone.

### What changes do you think we should make in order to solve the problem?

We can add the check to confirm if the user has right access to perform certain action.
With my understanding, I think the task creator and assignee can perform the action of Mark as Done or Mark As Incomplete.
For checkbox and button we can add the disable prop according to those user level permissions.
For menuItem we can add another condition in if block.

### What alternative solutions did you explore? (Optional)
None

**Comment by puneetlath** [comment]
@NicMendonca re-assigning over to you since i'm going OOO for the next week. Thanks!

**Comment by namhihi237** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The message “completed task” and “Re open task” does not appear in task detail when ticking the checkbox or press “Mark as done”, “Mark as incomplete” on task not assinged
### What is the root cause of that problem?
Now, Only the task creator and assignee can complete and repoen the task. After calling completeTask and reopenTask, reportAction is saved to Onyx with optimisticData, But after API return permission error, failureData has not deleted reportAction saved above.
https://github.com/Expensify/App/blob/4cad2489f6feb5efb955924764711a0c36439f58/src/libs/actions/Task.js#L204-L208
https://github.com/Expensify/App/blob/4cad2489f6feb5efb955924764711a0c36439f58/src/libs/actions/Task.js#L260-L264
### What changes do you think we should make in order to solve the problem?
To solve this problem we will set reportAction = null if the API returns an error.
In completedTask:
```
{
        onyxMethod: Onyx.METHOD.MERGE,
	key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${taskReportID}`,
	value: {[completedTaskReportAction.reportActionID]: null},
},
```
In reopenTask:
```
{
	onyxMethod: Onyx.METHOD.MERGE,
	key:`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${taskReportID}`,
	value: {[reopenedTaskReportAction.reportActionID]: null},
},
```
Result:

https://github.com/Expensify/App/assets/39086067/ab5eb069-f49e-4785-9020-5b12c01d20bf

### What alternative solutions did you explore? (Optional)
Disable button, checkbox where these 2 APIs are called by checking if current user is owner or assigned

**Comment by mollfpr** [comment]
Thanks for the proposals!

We will hide the **Mark as done** button in this issue https://github.com/Expensify/App/issues/19631#issuecomment-1582847844

So I'd like a solution for handling the failure action in this issue.

@alex-mechler I'm curious if we will have another error from the API other than the permission error on completing task?

**Comment by mollfpr** [comment]
@dukenv0307 Is the error from API is only about permission?

**Comment by alex-mechler** [comment]
> @alex-mechler I'm curious if we will have another error from the API other than the permission error on completing task?

I'm not quite sure I'm following here. Are you talking about other possible errors that can occur here?

**Comment by BhuvaneshPatil** [comment]
> Thanks for the proposals!
> 
> We will hide the **Mark as done** button in this issue [#19631 (comment)](https://github.com/Expensify/App/issues/19631#issuecomment-1582847844)
> 
> So I'd like a solution for handling the failure action in this issue.
> 
> @alex-mechler I'm curious if we will have another error from the API other than the permission error on completing task?

There will be three error - 
- Cancel
- Mark as done
- Mark as incomplete
 
 We will need to handle the `failureData` for all of them, I have updated my proposal which will satisfy the new needs
here - https://github.com/Expensify/App/issues/20542#issuecomment-1585007728

**Comment by mollfpr** [comment]
@alex-mechler Let's hold this for https://github.com/Expensify/App/pull/20689, where it handles the permission of the task action, and the action button probably will be hidden for non-assignee and not the task owner.

**Comment by alex-mechler** [comment]
That makes sense to me, placing this on hold

**Comment by NicMendonca** [comment]
Not deployed to staging just yet - https://github.com/Expensify/App/pull/20689#issuecomment-1596292401

**Comment by michaelhaxhiu** [comment]
🧇 I'm removing AlexM's assignment and triggering assignment of a new CME.

**Comment by NicMendonca** [comment]
This has been deployed: https://github.com/Expensify/App/pull/20689

@mollfpr @deetergp should we remove hold?  @namhihi237 is this still reproducible?

**Comment by deetergp** [comment]
@NicMendonca Yes, I think we can remove the hold, since that was the PR for the GH this is held on.

**Comment by mollfpr** [comment]
I will be checking this issue against the current `main`, and let you guys update soon.

**Comment by NicMendonca** [comment]
@mollfpr any update?

**Comment by mollfpr** [comment]
The issue is still reproducible when checking the checkbox by non-assignee users.

I suggest adding the error message when the user doesn't have permission to complete the task. Both proposals from @dukenv0307 and @BhuvaneshPatil are similar, but I recall that @dukenv0307 proposed it first. So, picking up their [proposal](https://github.com/Expensify/App/issues/20542#issuecomment-1586509453) to move forward.

🎀 👀 🎀 C+ reviewed!

**Comment by NicMendonca** [comment]
Removed Maria


@deetergp bump: https://github.com/Expensify/App/issues/20542#issuecomment-1615003766

**Comment by deetergp** [comment]
Yep, let's go with @dukenv0307's [proposal](https://github.com/Expensify/App/issues/20542#issuecomment-1586509453) 👍

**Comment by mollfpr** [comment]
![Screenshot 2023-07-03 at 22 16 33](https://github.com/Expensify/App/assets/25520267/1bf0abcb-1fd2-4106-8069-92cbe4a40ead)

We need help with the copy and the translation. Currently, @dukenv0307 is proposing `You do not have the permission to do the requested action.` for the error message when the user tries to complete an unassigned task. cc @NicMendonca @deetergp

**Comment by deetergp** [comment]
Oh, good question. That seems like a perfectly acceptable error message to me, but let's see what our Marketing folks have to say about it. I'm adding the `WaitingForCopy` label, which should assign someone from that team…

**Comment by deetergp** [comment]
A wild @shmaxey appears

**Comment by shmaxey** [comment]
I think that language makes sense 👍

**Comment by NicMendonca** [comment]
@mollfpr just to confirm you'll be requesting payment via Expensify?

**Comment by mollfpr** [comment]
@NicMendonca Nope, still using same old Expensify.

**Comment by NicMendonca** [comment]
everyone has been paid ✅ BZ checklist done ✅ 

Thanks!

**Comment by NicMendonca** [comment]
@dukenv0307 I am not reading this as within 3 business days: 

- You were assigned on June 30th  
- PR Opened July 2nd
- C+ Approved changed ready for internal Eng on July 9th

**Comment by NicMendonca** [comment]
> We request the internal engineer for the first time on July 6th

I'm sorry - I am still not reading this as the **PR being merged within 3 business days**. 

If you were assigned on the 30th on June and requested internal eng on July 6th, then that is 5 business days

**Comment by NicMendonca** [comment]
discussing internally [here](https://expensify.slack.com/archives/C01SKUP7QR0/p1690758362569699)

**Comment by NicMendonca** [comment]
@dukenv0307 I am not counting the weekend. 

I am counting as followed: 

- Friday June 30th - you were assigned  - this is the start of the timer
- Monday July 3rd - day 2 
- Tuesday July 4th - day 3
- Wednesday July 5th - day 5 

^ This is 4 business days.
